vmsize.go

```diff
@@ -0,0 +1,156 @@
+	if vmSizeName == "" {
+		return nil, fmt.Errorf("error parsing VM size, it is empty")
+	}
+
+	parts := skuSizeScheme.FindStringSubmatch(vmSizeName)
+
+	if parts == nil || len(parts) < 8 {
+		return nil, fmt.Errorf("could not parse VM size %s", vmSizeName)
+	}
+
+	result := make(map[string]string)
+	for part, index := range partIndexes {
+		result[part] = parts[index]
+	}
+	return result, nil
+}
+
+// GetSubFamily fetches the 'subFamily' part from the SKU size and returns it.
+func (s *SKU) GetSubFamily() (*string, error) {
+	return getSubFamily(s.GetSize())
+}
+
+func getSubFamily(vmSizeName string) (*string, error) {
+	parts, err := parseVMSize(vmSizeName)
+	if err != nil {
+		return nil, err
+	}
+	if len(parts[SubFamilyKey]) > 0 {
+		subfamily := parts[SubFamilyKey]
+		return &subfamily, nil
+	}
+	return nil, nil
+}
+
+// GetConstrainedVCPUs fetches the 'constrainedVCPUs' part from the SKU size and returns it.
+func (s *SKU) GetConstrainedVCPUs() (*string, error) {
+	return getConstrainedVCPUs(s.GetSize())
+}
+
+func getConstrainedVCPUs(vmSizeName string) (*string, error) {
+	parts, err := parseVMSize(vmSizeName)
+	if err != nil {
+		return nil, err
+	}
+
+	constrainedVCPUs := parts[ConstrainedVCPUsKey]
+	if len(constrainedVCPUs) > 0 {
+		_, err := strconv.Atoi(constrainedVCPUs)
+		if err != nil {
+			return nil, fmt.Errorf("error converting constrained VCPUs, %w", err)
+		}
+		return &constrainedVCPUs, nil
+	}
+	return nil, nil
+}
+
+// GetAdditiveFeatures fetches the 'additiveFeatures' part from the SKU size and returns it.
+func (s *SKU) GetAdditiveFeatures() ([]rune, error) {
+	return getAdditiveFeatures(s.GetSize())
+}
+
+func getAdditiveFeatures(vmSizeName string) ([]rune, error) {
+	parts, err := parseVMSize(vmSizeName)
+	if err != nil {
+		return nil, err
+	}
+	return []rune(parts[AdditiveFeaturesKey]), nil
+}
+
+// GetAcceleratorType fetches the 'acceleratorType' part from the SKU size and returns it.
+func (s *SKU) GetAcceleratorType() (*string, error) {
+	return getAcceleratorType(s.GetSize())
+}
+
+func getAcceleratorType(vmSizeName string) (*string, error) {
+	parts, err := parseVMSize(vmSizeName)
+	if err != nil {
+		return nil, err
+	}
+	if len(parts[AcceleratorTypeKey]) > 0 {
+		acceleratorType := parts[AcceleratorTypeKey]
+		return &acceleratorType, nil
+	}
+	return nil, nil
+}
+
+// GetVersion fetches the 'version' part from the SKU size and returns it.
+func (s *SKU) GetVersion() (string, error) {
+	return getVersion(s.GetSize())
+}
+
+func getVersion(vmSizeName string) (string, error) {
+	parts, err := parseVMSize(vmSizeName)
+	if err != nil {
+		return "", err
+	}
+	return parts[VersionKey], nil
+}
+
+func (s *SKU) GetSeries() (string, error) {
+	return getSeries(s.GetSize(), s.GetFamilyName())
+}
+
+func getSeries(vmSizeName, family string) (string, error) {
+	subFamilyPtr, err := getSubFamily(vmSizeName)
+	subFamily := ""
+	if subFamilyPtr != nil {
+		subFamily = *subFamilyPtr
+	}
```

Comment:
maybe FromPtr from the lo or the azure sdk for go to equiv https://pkg.go.dev/github.com/samber/lo#FromPtr instead